diff -ru yum-3.2.22-orig/yum/__init__.py yum-3.2.22/yum/__init__.py
--- yum-3.2.22-orig/yum/__init__.py	2010-02-23 10:29:42.000000000 -0500
+++ yum-3.2.22/yum/__init__.py	2010-02-23 10:30:31.000000000 -0500
@@ -561,7 +561,7 @@
         
         if self.conf.obsoletes:
             obs_init = time.time()    
-            self._up.rawobsoletes = self.pkgSack.returnObsoletes()
+            self._up.rawobsoletes = self.pkgSack.returnObsoletes(newest=True)
             self.verbose_logger.debug('up:Obs Init time: %0.3f' % (time.time() - obs_init))
             
         self._up.exactarch = self.conf.exactarch
Only in yum-3.2.22/yum: __init__.py~
